src/internal/kopia/base_finder.go

```diff
@@ -0,0 +1,259 @@
+	"github.com/alcionai/corso/src/internal/operations/inject"
+	"github.com/alcionai/corso/src/pkg/backup"
+	"github.com/alcionai/corso/src/pkg/logger"
+)
+
+type BackupBases struct {
+	Backups     []BackupEntry
+	MergeBases  []ManifestEntry
+	AssistBases []ManifestEntry
+}
+
+type BackupEntry struct {
+	*backup.Backup
+	Reasons []Reason
+}
+
+type baseFinder struct {
+	sm snapshotManager
+	bg inject.GetBackuper
+}
+
+func NewBaseFinder(
+	sm snapshotManager,
+	bg inject.GetBackuper,
+) (*baseFinder, error) {
+	if sm == nil {
+		return nil, clues.New("nil snapshotManager")
+	}
+
+	if bg == nil {
+		return nil, clues.New("nil GetBackuper")
+	}
+
+	return &baseFinder{
+		sm: sm,
+		bg: bg,
+	}, nil
+}
+
+func (b *baseFinder) getBackupModel(
+	ctx context.Context,
+	man *snapshot.Manifest,
+) (*backup.Backup, error) {
+	k, _ := makeTagKV(TagBackupID)
+	bID := man.Tags[k]
+
+	ctx = clues.Add(ctx, "search_backup_id", bID)
+
+	bup, err := b.bg.GetBackup(ctx, model.StableID(bID))
+	if err != nil {
+		return nil, clues.Stack(err).WithClues(ctx)
+	}
+
+	return bup, nil
+}
+
+// findBasesInSet goes through manifest metadata entries and sees if they're
+// incomplete or not. If an entry is incomplete and we don't already have a
+// complete or incomplete manifest add it to the set for kopia assisted
+// incrementals. If it's complete, fetch the backup model and see if it
+// corresponds to a successful backup. If it does, return it as we only need the
+// most recent complete backup as the base.
+func (b *baseFinder) findBasesInSet(
+	ctx context.Context,
+	reason Reason,
+	metas []*manifest.EntryMetadata,
+) (*BackupEntry, *ManifestEntry, []ManifestEntry, error) {
+	// Sort manifests by time so we can go through them sequentially. The code in
+	// kopia appears to sort them already, but add sorting here just so we're not
+	// reliant on undocumented behavior.
+	sort.Slice(metas, func(i, j int) bool {
+		return metas[i].ModTime.Before(metas[j].ModTime)
+	})
+
+	var (
+		kopiaAssistSnaps []ManifestEntry
+		foundIncomplete  bool
+	)
+
+	for i := len(metas) - 1; i >= 0; i-- {
+		meta := metas[i]
+		ictx := clues.Add(ctx, "search_snapshot_id", meta.ID)
+
+		man, err := b.sm.LoadSnapshot(ictx, meta.ID)
+		if err != nil {
+			// Safe to continue here as we'll just end up attempting to use an older
+			// backup as the base.
+			logger.CtxErr(ictx, err).Info("attempting to get snapshot")
+			continue
+		}
+
+		if len(man.IncompleteReason) > 0 {
+			if !foundIncomplete {
+				foundIncomplete = true
+
+				kopiaAssistSnaps = append(kopiaAssistSnaps, ManifestEntry{
+					Manifest: man,
+					Reasons:  []Reason{reason},
+				})
+			}
+
+			continue
+		}
+
+		// This is a complete snapshot so see if we have a backup model for it.
+		bup, err := b.getBackupModel(ictx, man)
+		if err != nil {
+			if errors.Is(err, data.ErrNotFound) {
+				continue
+			}
+
+			// Safe to continue here as we'll just end up attempting to use an older
```

Comment:
Double checking: the base backup defines our delta token set, right?

Reply:
in this context yes. The code is a little loose on terminology, but a "valid base backup" is actually a backup with the following:
* complete item data snapshot
* backup model
* populated SnapshotID in backup model
* populated details (StreamStoreID/DetailsID) in backup model